### PR TITLE
Docs: Autogenerate API reference content

### PR DESCRIPTION
**Pull request by oraNod**
This PR uses the immaterial extension with Sphinx to generate API docs.

To do:
- Find a fix for the warnings that are ignored in by nitpicky in conf.py
- Fix additional warnings during tox build.

### CODE REVIEW

File: docs/source/conf.py [hunk 1]
```diff
@@ -15,24 +15,42 @@
...
+}
 
-source_suffix = {
-    ".rst": "restructuredtext",
-    ".md": "markdown",
+intersphinx_mapping = {
+    'python': ('https://docs.python.org/3', None),
 }
 
+# Temporarily ignore warnings with apidoc generation.
```
Comment (webknjaz) [inline docs/source/conf.py:33]:
```suggestion
# Temporarily ignore "missing reference" warnings with apidoc generation.
```

File: docs/source/conf.py [hunk 2]
```diff
@@ -15,24 +15,42 @@
...
+    ('py:class', 'ansible_sdk.executors.subprocess._AnsibleContainerJobExecutorBase'),
+    ('py:class', 'ansible_sdk.executors.subprocess._AnsibleContainerJobOptions'),
+    ('py:class', 'ansible_sdk._util.dataclass_compat._DataclassReplaceMixin'),
+    ('py:class', 'ansible_sdk.executors.subprocess._AnsibleContainerJobExecutorBase'),
+    ('py:class', 'ansible_sdk.executors.subprocess.OptionsT'),
+    ('py:class', 'ansible_sdk.executors.subprocess._AnsibleContainerJobOptions'),
+    ('py:class', 'ansible_sdk.executors.mesh.AnsibleMeshJobOptions'),
+]
+
+templates_path = ['_templates']
```
Comment (webknjaz) [inline docs/source/conf.py:51]: This is unnecessary unless you're going to integrate custom template overrides placed under that directory.

File: docs/Makefile [hunk 3]
```diff
@@ -1,20 +0,0 @@
-# Minimal makefile for Sphinx documentation
```
Comment (webknjaz) [inline docs/Makefile:1]: I'd just put `Makefile` and `make.bat` removals in a separate PR.

File: .readthedocs.yaml [hunk 4]
```diff
@@ -9,10 +9,6 @@ build:
   os: ubuntu-20.04
   tools:
     python: "3.9"
-    # You can also specify other tool versions:
```
Comment (webknjaz) [inline .readthedocs.yaml:12]: How about moving potential changes to the RTD integration to a separate PR so that the results of several unrelated changes don't create confusion during debugging?

File: docs/doc-requirements.txt [hunk 5]
```diff
@@ -1,3 +1,4 @@
 Sphinx==5.1.1
 sphinx-ansible-theme==0.9.1
+sphinx-immaterial==0.9.1
 sphinxcontrib-apidoc==0.3.0
```
Comment (webknjaz) [inline docs/doc-requirements.txt:4]: Since this PR replaces the use of this extension, it no longer has to be installed.
```suggestion
```

File: docs/source/conf.py [hunk 6]
```diff
@@ -15,24 +15,42 @@
 # https://www.sphinx-doc.org/en/master/usage/configuration.html#general-configuration
 
 extensions = [
+    'sphinx.ext.intersphinx',
     'sphinx_ansible_theme',
-    'sphinxcontrib.apidoc'
+    'sphinx_immaterial',
```
Comment (webknjaz) [inline docs/source/conf.py:20]: Looks like this extension is auto-injecting their integrated theme which is what's probably causing those them option warnings:
```console
WARNING: unsupported theme option 'font' given
WARNING: unsupported theme option 'plugins' given
WARNING: unsupported theme option 'icon' given
WARNING: unsupported theme option 'repo_url' given
WARNING: unsupported theme option 'edit_uri' given
WARNING: unsupported theme option 'toc_title' given
WARNING: unsupported theme option 'toc_title_is_page_title' given
```
_(copied from the CI output)_

It is probably doing too much and maybe it's possible to only initialize the necessary bits by loading separate sub-extensions, I'm not sure but it's worth a try.

Comment (webknjaz) [reply docs/source/conf.py:20]: https://jbms.github.io/sphinx-immaterial/apidoc/python/apigen.html#usage doesn't say that adding `sphinx_immaterial` is required there. So maybe dropping this would fix some problems:
```suggestion
```

Comment (oraNod) [reply docs/source/conf.py:19]: Dropping 'spinx_immaterial' leads to errors such as:

```
Sphinx parallel build error:
ValueError: Failed to document entity: 'ansible_sdk.AnsibleJobDef'
```
I'd like to see if there is another way to resolve those unsupported theme option warnings.

**Comment by webknjaz** [comment]
> ```console
> /home/runner/work/ansible-sdk/ansible-sdk/ansible_sdk.AnsibleJobDef:1: WARNING: Parameter name 'data_dir' does not match any of the parameters defined in the signature: []
> ```

Things like this seem to be happening to data classes where the `:param:` entries are declared in the class docstring, not in `__init__()` (which is not there). The docs don't mention anything special about the dataclasses, nor do the GitHub issues. I suppose one could file a feature request to infer an initializer signature from the attributes but currently, the workaround could be just documenting those attributes.
For example, https://github.com/ansible/ansible-sdk/blob/main/ansible_sdk/model/job_def.py could have a class changed as follows:
```diff
 @dataclass(frozen=True, kw_only=True)
 class AnsibleJobDef(_DataclassReplaceMixin):
-    """
-    Common Ansible job definition values
-    :param data_dir: path to a directory structure containing the Ansible project content and inventory.
-    :param playbook: relative path or FQCN of a playbook to run under ``data_dir``
-    :param inventory: relative path to inventory file(s) or directory under ``data_dir``, analogue to Ansible's ``-i`` option
-    :param extra_vars: dictionary of variables for highest precedence level, analogue to Ansible's ``-e`` option
-    :param verbosity: None for default verbosity or 1-5, equivalent to Ansible's ``-v`` options
-     """
+    """Common Ansible job definition values."""
     # currently analogue to runner's private_data_dir; do we want to construct this ourselves?
     data_dir: str
+   """Path to a directory structure containing the Ansible project content and inventory."""
+
     # relative path to playbook in data_dir or FQCN
     playbook: str
+   """Relative path or FQCN of a playbook to run under ``data_dir``."""
+
     inventory: t.Optional[t.Union[str, list[str]]] = None  # FUTURE: high-level inventory types?
+   """Relative path to inventory file(s) or directory under ``data_dir``, analogue to Ansible's ``-i`` option."""
+
     extra_vars: dict[str, t.Any] = field(default_factory=dict)
+   """Dictionary of variables for highest precedence level, analogue to Ansible's ``-e`` option."""
+
     verbosity: t.Optional[int] = None  # None or 1-5
+   """:data:`None` for default verbosity or 1-5, equivalent to Ansible's ``-v`` options."""
```

The above example uses attribute docstrings (check out PEP 224, PEP 257, PEP 258 for the origins of this approach). Sphinx's apidoc supports them natively.

But there's another syntax that Sphinx supports for documenting attributes — comment-like strings that have a leading `#: ` and are supposed to be on the previous or the same line, unlike attribute docstrings that go on the line following the attribute declaration (https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html#directive-autoattribute).
Here's an example of using this syntax instead:

```diff
 @dataclass(frozen=True, kw_only=True)
 class AnsibleJobDef(_DataclassReplaceMixin):
-    """
-    Common Ansible job definition values
-    :param data_dir: path to a directory structure containing the Ansible project content and inventory.
-    :param playbook: relative path or FQCN of a playbook to run under ``data_dir``
-    :param inventory: relative path to inventory file(s) or directory under ``data_dir``, analogue to Ansible's ``-i`` option
-    :param extra_vars: dictionary of variables for highest precedence level, analogue to Ansible's ``-e`` option
-    :param verbosity: None for default verbosity or 1-5, equivalent to Ansible's ``-v`` options
-     """
+    """Common Ansible job definition values."""
     # currently analogue to runner's private_data_dir; do we want to construct this ourselves?
+   #: Path to a directory structure containing the Ansible project content and inventory
     data_dir: str
+
     # relative path to playbook in data_dir or FQCN
+   #: Relative path or FQCN of a playbook to run under ``data_dir``
     playbook: str
+
+   #: Relative path to inventory file(s) or directory under ``data_dir``, analogue to Ansible's ``-i`` option
     inventory: t.Optional[t.Union[str, list[str]]] = None  # FUTURE: high-level inventory types?
+
+   #: Dictionary of variables for highest precedence level, analogue to Ansible's ``-e`` option
     extra_vars: dict[str, t.Any] = field(default_factory=dict)
+
+   #: :data:`None` for default verbosity or 1-5, equivalent to Ansible's ``-v`` options
     verbosity: t.Optional[int] = None  # None or 1-5
```

I haven't tested these but I'm pretty confident that such a workaround should address the corresponding warnings.

File: docs/source/conf.py [hunk 7]
```diff
@@ -15,24 +15,42 @@
 # https://www.sphinx-doc.org/en/master/usage/configuration.html#general-configuration
 
 extensions = [
+    'sphinx.ext.intersphinx',
```
[the remark this answers is not in the document]
Comment (oraNod) [reply docs/source/conf.py]: Good suggestion, thanks. https://github.com/ansible/ansible-sdk/pull/42

File: docs/Makefile (hunk 3 above)
Comment (oraNod) [reply docs/Makefile]: Done with https://github.com/ansible/ansible-sdk/pull/41

File: docs/source/conf.py (hunk 6 above)
Comment (webknjaz) [reply docs/source/conf.py:19]: That error is coming from https://github.com/jbms/sphinx-immaterial/blob/21f2aeeb7fd182955a78893c30aaa6b84d6d5e8e/sphinx_immaterial/apidoc/python/apigen.py#L628.

I'd maybe try having a simple `setup()` function in this `conf.py` with just these: https://github.com/jbms/sphinx-immaterial/blob/a944f27/sphinx_immaterial/__init__.py#L254-L267 to see if it initializes enough internals for this to work.

File: tox.ini [hunk 8]
```diff
@@ -22,9 +22,7 @@ commands = pytest {posargs:tests/unit}
 [testenv:docs]
 description = Build documentation
 deps = -r{toxinidir}/docs/doc-requirements.txt
-commands = 
-  sphinx-apidoc -o docs/source/api ansible_sdk
-  sphinx-build -T -E -n --keep-going {tty:--color} -j auto -d docs/build/doctrees -b html docs/source docs/build/html
+commands = sphinx-build -T -E -W -n --keep-going {tty:--color} -j auto -d {temp_dir}/doctrees -b html docs/source docs/build/html
```
Comment (webknjaz) [inline tox.ini:25]: (nit) I'd still keep it on a separate line for the sake of nice diffs and fitting things on screen.

Also, the `-W` and `{temp_dir}` could use a separate PR. Adding them is not related to the immaterial ext.

Comment (oraNod) [reply tox.ini:25]: Done with https://github.com/ansible/ansible-sdk/pull/43

**Comment by nitzmahone** [comment]
I haven't done an in-depth review of these changes, but looking at the resultant output gets us back to where we need to be, so I'm a big +1 for the approach.

**Review by nitzmahone — COMMENTED** [review]
I haven't done an in-depth review of these changes, but looking at the resultant output gets us back to where we need to be, so I'm a big +1 for the approach.

**Comment by nitzmahone** [comment]
> But there's another syntax that Sphinx supports for documenting attributes — comment-like strings that have a leading #: 

@webknjaz thanks for that one- I didn't do my research to find this syntax when I added those docstrings, but it sounds like exactly what I was wishing for. I'd much prefer the docstrings to be attribute/property-adjacent like that than having them in a block at the top. We don't need to address it in this PR, but I'd be strongly in favor of doing it that way if it doesn't cause other problems.

**Comment by webknjaz** [comment]
FWIW I personally tend to prefer attribute docstrings rather than the Sphinx hack with `#:`.

> We don't need to address it in this PR

If the warnings were caused by this change, it'd be reasonable to fix it here. If this is also happening with the current apidoc integration, then definitely in a separate PR. I haven't checked if they're already present on the default branch.

**Comment by oraNod** [comment]
Update on this PR. I propose that we create two conf.py files and have separate builds. One build for concepts and procedures (docs/source/conf.py). One build for API reference generated from docstrings (docs/api_reference/conf.py). IMO a separate build for the API reference content will make it easier to have a single-source for community and downstream.

I also propose that we fix the remaining warnings in a different PR because that requires changes to the docstrings.

```
/home/dnaro/git/dnaro-sdk-build-api-docs/ansible_sdk.AnsibleJobDef:1: WARNING: Parameter name 'data_dir' does not match any of the parameters defined in the signature: []
/home/dnaro/git/dnaro-sdk-build-api-docs/ansible_sdk.AnsibleJobDef:1: WARNING: Parameter name 'playbook' does not match any of the parameters defined in the signature: []
/home/dnaro/git/dnaro-sdk-build-api-docs/ansible_sdk.AnsibleJobDef:1: WARNING: Parameter name 'inventory' does not match any of the parameters defined in the signature: []
/home/dnaro/git/dnaro-sdk-build-api-docs/ansible_sdk.AnsibleJobDef:1: WARNING: Parameter name 'extra_vars' does not match any of the parameters defined in the signature: []
/home/dnaro/git/dnaro-sdk-build-api-docs/ansible_sdk.AnsibleJobDef:1: WARNING: Parameter name 'verbosity' does not match any of the parameters defined in the signature: []
/home/dnaro/git/dnaro-sdk-build-api-docs/ansible_sdk.executors.AnsibleDockerJobOptions:1: WARNING: Parameter name 'container_image_ref' does not match any of the parameters defined in the signature: []
/home/dnaro/git/dnaro-sdk-build-api-docs/ansible_sdk.AnsibleJobStatusEvent:1: WARNING: Parameter name 'event' does not match any of the parameters defined in the signature: []
/home/dnaro/git/dnaro-sdk-build-api-docs/ansible_sdk.AnsibleJobStatusEvent:1: WARNING: Parameter name 'event_data' does not match any of the parameters defined in the signature: []
/home/dnaro/git/dnaro-sdk-build-api-docs/ansible_sdk.AnsibleJobStatusEvent:1: WARNING: Parameter name 'stdout' does not match any of the parameters defined in the signature: []
/home/dnaro/git/dnaro-sdk-build-api-docs/ansible_sdk.executors.AnsiblePodmanJobOptions:1: WARNING: Parameter name 'container_image_ref' does not match any of the parameters defined in the signature: []
```
